### PR TITLE
Example comparing`cheri_cap_build` & `cheri_address_set`

### PR DESCRIPTION
**Pull request by probablytom**
`cheri_cap_build` and `cheri_address_set` are similar, but not quite the same. We were previously unsure of what the difference between the two was; this example uses both and explains the differences between the two.

It's pretty explicit about what's going on and prints detailled messages describing the capabilities it constructs & prints. I figured it'd be useful to have some detail in there so we didn't have to re-parse the code every time we wanted to remember what the difference between the functions is.

Put together mostly by @0152la; I just cut down the program we were using as a scratchpad when figuring this out and tidied it into an example.

### CODE REVIEW

File: cap_build.c
Comment (0152la) [inline cap_build.c:22]: I would change these prints into comments, so on executing, we just have the capability prints, and then can refer to side-by-side with the example code. It's more that this is how all examples (to my knowledge) are done.

File: cap_build.c
Comment (0152la) [inline cap_build.c:42]: Can you explain why we need the `uintcap_t` cast here? I *think* it's because it's 128 bits, rather than a capability's 129, but not sure. You can also just say we need this since it works.

File: cap_build.c
Comment (0152la) [inline cap_build.c:63]: I would either drop this whole discussion (or sprinkle it in comments appropriately), or have an example which emphasises the difference (which the rest of this example might do).

**Review by 0152la — CHANGES_REQUESTED** [review]

**Comment by ltratt** [comment]
Since I'm the dummy here, I'll ask the obvious question: does `cheri_cap_build` allow you to create an arbitrary capability that overrides all other checks? I guess "no", but I'm not quite sure what the "but" is. My initial guess is that the capability that's created is valid but has limited permissions -- but maybe I'm wrong!

**Comment by ltratt** [comment]
Thanks!

> I have a suspicion it can be used to reconstruct a capability that's been serialised to a file

This shouldn't be possible -- *but* there was once a bug in CheriBSD that allowed it because of file caching! It was fixed, though.

**Comment by djlowther** [comment]
`cheri_cap_build` translates to the `BUILD` Morello instruction, which like Tom suspects turns a bit-pattern into a valid, unsealed capability, provided that that capability can be derived from the capability argument. I believe it's designed for situations like virtual memory paging -- the kernel can use it to quickly reconstitute capabilities in swapped pages rather than having to use a sequence of SCBNDS, SCVALUE, etc.

**Comment by ltratt** [comment]
Thanks both -- I get it now!

File: cap_build.c
Comment (probablytom) [reply cap_build.c]: Sorry for the delay. Fixed in [be6d139](https://github.com/capablevms/cheri-examples/pull/92/commits/be6d139256ad63fa4067d244d8082f8926efe9fd) — thanks!

File: cap_build.c
Comment (probablytom) [reply cap_build.c:59]: We need this since it works 😛

More seriously though, the function only accepts a `uintcap_t`. You might remember when we were pairing on this that we got lots of cryptic errors from the compiler around the type of this argument. It wanted a `uintcap_t` — I imagine because it specifically doesn't care about the validity bit of its second argument (as it'll decide for itself whether the capability it returns should be valid), but it needs every other bit of the capability, because whatever it returns will be a capability with these 128 bits…just not necessarily a valid one.

File: cap_build.c
Comment (probablytom) [reply cap_build.c]: I reworked it (with updates following @djlowther's explanation of `cheri_cap_build` from today — thanks!) and moved the relevant parts to the top of the example as an explanation of what the example is trying to demonstrate / accomplish and as a brief intro for anybody using it to learn. WDYT of the current incarnation? (Fixed in [a1b2ba1](https://github.com/capablevms/cheri-examples/pull/92/commits/a1b2ba1cc12effdf2fa5b0bce7d8ee3bf7b6ddd8))

**Comment by probablytom** [comment]
Thanks @djlowther — that's a very helpful note! We haven't found much documentation around `cheri_cap_build` ourselves.

I haven't played with this yet and can't budget time to investigating it right now, but perhaps you know already: does that mean that a capability written to disk by one process could be loaded by a hybrid process with a very permissive DDC, and so re-created as a valid capability by that second process? Shouldn't that be forbidden, as @ltratt mentioned? Or am I misunderstanding something?

**Comment by probablytom** [comment]
Fascinating! Thank you for the detailled response!

File: cap_build.c
Comment (0152la) [reply cap_build.c:59]: I meant more in the code, as a comment. I recall we found some builtin function signature of this `cheri_cap_build` that took two identically typed arguments in the representation, but passing a capability as the second one gave an error (while obviously a capability as the first one is fine). 

If you, by chance, found some more details about the function signature of this, that would be great, if not, no worries. Can say we experimented and this seems to be the signature, just add that as a comment.

Comment (probablytom) [reply cap_build.c:64]: Ah! Sorry, misunderstood. Closest to a definition I ever found was at the end of [this article from ARM](https://www.morello-project.org/resources/morello-linux-morelloie/). Added a note on that in [f0dbba6](https://github.com/capablevms/cheri-examples/pull/92/commits/f0dbba61699584ac1fa59191d2d7c276e2662669).

Comment (0152la) [reply cap_build.c:64]: Right, Yury did mention when that article got uploaded. I wonder how related it is to his examples repo.

**Comment by ltratt** [comment]
@0152la I think this is OK to squash?

**Review by 0152la — APPROVED** [review]

**Comment by probablytom** [comment]
Squashed!

**Comment by 0152la** [comment]
On my end, still says it's out of date with master.

**Comment by probablytom** [comment]
Oh, and mine too. Apologies, thought I'd fixed — one sec

**Comment by 0152la** [comment]
I've removed it from the merge queue, since this now seems to have our whole repo history in it.

**Comment by probablytom** [comment]
Yes, fortunately it looks like it was also still marked as out of sync with master so didn't auto-merge. Should be fixed now — looks correct on my end, at least — but will keep my mitts off the merge button…

**Comment by 0152la** [comment]
Forgot one thing; should we add this to CI? I would, but I'll defer to @ltratt if we want to merge as is, since it's more "this is how you use this" of an example.

**Comment by ltratt** [comment]
Good point about adding it to CI. @probablytom Can you push a commit which does that please?

**Comment by probablytom** [comment]
Hang on while I rebase again to resolve the conflict…

**Comment by 0152la** [comment]
LGTM

**Comment by ltratt** [comment]
Please squash.

**Comment by probablytom** [comment]
Sorry for the noise. Should be good now.

File: tests/run_tests.sh
Comment (ltratt) [inline tests/run_tests.sh:53]: Merge conflict left in. Please force push an update.

Comment (probablytom) [reply tests/run_tests.sh:53]: 🤦

Comment (probablytom) [reply tests/run_tests.sh]: Fixed in force-pushed [f509753](https://github.com/capablevms/cheri-examples/pull/92/commits/f509753a0b6fa89ee15be5499c747ccb018fb704).

**Comment by ltratt** [comment]
Please squash.

**Comment by probablytom** [comment]
…should be good to go… 👍

**Comment by probablytom** [comment]
Re-formatted.

**Comment by ltratt** [comment]
@probablytom Is this ready for rereview?

**Comment by 0152la** [comment]
I'm confused about commit https://github.com/capablevms/cheri-examples/pull/92/commits/ba68b10ed686829e326fb955857718d4cfa53486. Looking at the code [1], this new example should only run on `morello-hybrid`. If you see it running on `risc-v`, that might be a CI issue, and I need to investigate.

[1] https://github.com/capablevms/cheri-examples/pull/92/files#diff-bfa67bbd428f0fd40485bf6e666132a2adb30f6524631aff49d2bcd30f71c5ebL75

**Comment by ltratt** [comment]
>

**Comment by ltratt** [comment]
> I'm confused about commit https://github.com/capablevms/cheri-examples/commit/ba68b10ed686829e326fb955857718d4cfa53486. Looking at the code [1], this new example should only run on morello-hybrid. If you see it running on risc-v, that might be a CI issue, and I need to investigate.

I agree that this might be worth investigating, but I'd prefer to get this PR in now, and investigate separately: we can always reenable on RISC-V later if it turns out there's a CI issue.

@probablytom Just checking that this PR is finished from your perspective?

**Comment by probablytom** [comment]
Yes, I was confused as to why it was running too. I think it's because the CI runs `make -f Makefile.riscv64-purecap all`, and the make rule appears to pick up _every_ example in the root of the repo and run it as if it were riscv-compatible. Other examples have similar macros to avoid running on the incorrect architectures [1], I imagine for exactly this reason — so I _think_ that commit [2] actually follows the precedent set by other examples so far. Afraid I haven't been able to investigate further, but I'm confident the change should fix the issue.

Given other examples do this too, if the build passes CI now I don't think it's worth making changes in the short term to get this to run without the macros…but if you're keen, my suspicions sit with that makefile.

So to answer the Q @ltratt, yes — happy with the state of this now — but also happy to make changes if there's anything else to tidy up! I'll squash when instructed.

[1] https://github.com/capablevms/cheri-examples/blob/master/general_bounds.c
[2] https://github.com/capablevms/cheri-examples/commit/ba68b10ed686829e326fb955857718d4cfa53486

**Comment by 0152la** [comment]
You're right. It seems it assumes root examples mean `purecap` [1]. Looking at the logic then [2], you should place this example in `./hybrid/`, and remove the additional condition you added in `run_tests.sh`.

I *think* that's the best resolution, but still the fact that we have some tests compiling in `.buildbot.sh`, then others running and compiling in `./tests/run_tests.sh` is not ideal, and should probably fix that. However, that is beyond the scope of this PR. If this works and @ltratt is fine merging as is, without the change at the top, I'm fine with that.

[1] https://github.com/capablevms/cheri-examples/blob/master/.buildbot.sh#L18
[2] https://github.com/capablevms/cheri-examples/blob/master/.buildbot.sh#L29

**Comment by ltratt** [comment]
Yeah, let's get this in, and then refine/move things around in another PR.

@probablytom Please squash.

**Comment by probablytom** [comment]
Squashed!

**Comment by ltratt** [comment]
Please squash.

**Comment by probablytom** [comment]
Squashed.

**Pull request merged**
